server/src/sim.rs

```diff
@@ -242,42 +254,69 @@ impl Sim {
     fn snapshot_node(&self, node: NodeId) -> save::EntityNode {
         let mut entities = Vec::new();
         for &entity in self.graph_entities.get(node) {
-            let Ok(entity) = self.world.entity(entity) else {
-                error!("stale graph entity {:?}", entity);
-                continue;
-            };
-            let Some(id) = entity.get::<&EntityId>() else {
-                continue;
-            };
-            let mut components = Vec::new();
-            if let Some(pos) = entity.get::<&Position>() {
-                components.push((
-                    ComponentType::Position as u64,
-                    postcard::to_stdvec(&pos.local.as_ref()).unwrap(),
-                ));
-            }
-            if let Some(ch) = entity.get::<&Character>().or_else(|| {
-                entity
-                    .get::<&InactiveCharacter>()
-                    .map(|ich| hecs::Ref::map(ich, |ich| &ich.0)) // Extract Ref<Character> from Ref<InactiveCharacter>
-            }) {
-                components.push((ComponentType::Name as u64, ch.name.as_bytes().into()));
-            }
-            let mut repr = Vec::new();
-            postcard_helpers::serialize(
-                &SaveEntity {
-                    entity: id.to_bits().to_le_bytes(),
-                    components,
-                },
-                &mut repr,
-            )
-            .unwrap();
-            entities.push(repr);
+            let reprs = self.snapshot_entity_and_children(entity);
```

Comment:
Are child entities not also in `graph_entities`?

Reply:
Not currently, although given how tricky this implementation is, I would consider this to be a mistake on my part.

There's a distinction between entities that have an inherent position (which is managed with a `Position` component) and nodes that don't, such as inventory items. For the latter, they belong to an entity that does have a position.

I had decided that `graph_entities` would only contain entities that have a position, but I have a good reason to second-guess that, since that doesn't correspond with how save files are organized at all. I'll go ahead and fix this before my next push to this branch.

Reply:
I've updated the PR so that child entities are now in `graph_entities`. To accomplish this (what I believe to be) the most understandable way possible, I've made `NodeId` into its own component intrinsically linked with `graph_entities` so that inventory items can be associated with a `NodeId` without needing a `Position`.

This is helpful because `Position` has a bunch of other functionality, such as syncing with clients, so separating this functionality is important.